src/goto-symex/goto_symex.cpp

```diff
@@ -114,9 +114,27 @@ void goto_symext::symex_assign(
         return;
     }
 
+    // We may end up reading (and writing) all components of an object composed
+    // of multiple fields. In such cases, we must do so atomically to avoid
+    // overwriting components modified by another thread. Note that this also
+    // implies multiple shared reads on the rhs being treated as atomic.
```

Comment:
Thanks for the comment that makes this clear but ... I am a little concerned about this because now
`struct s = t;` and

```
struct s;
s.x = t.x;
s.y = t.y;
```

can have different behaviour.  I guess there is a very valid question about what the `atomic` actions in the CPROVER concurrency model are.

Reply:
Although this case shouldn't arise for structs and mainly affects arrays: you are absolutely right that this is questionable either way. Without this patch, we'll have spurious results (as documented in the regression test). With this patch, we may miss some behaviour, depending on what exactly we otherwise consider to be atomic.